luci-lib-jsonc: Add ltn12-compatible sink factory

jplitza opened the pull request.

hnyman commented:
You should sign your commits.
(good practice is to setup git to have a commit message template with the signed-off-by line ready there.

jow- commented:
Can you add some LuaDoc documentation as well?

jplitza commented:
@hnyman I don't understand or like the practice of those lines, but of course will add one for these commits. Sorry I only now realized you have contribution guidelines demanding that, because I open the PR via API.

@jow- I didn't write any LuaDoc so far, but surely will try. However, the present luci-lib-jsonc LuaDoc doesn't really compile well for me (except for files, nothing is documented at all in the result, even though the source contains detailed descriptions of the functions).

On another note, @NeoRaider commented in the related issue freifunk-gluon/gluon#470 

> I find the way the sink is attached to the module quite awkward (and making C metatables accessible always asks for trouble).
>
> Just having a normal function that can be used like sink(decoder) instead of decoder:sink() seems much cleaner to me.

I would like to discuss those issues here before merging. I was guided by the way `nixio` does things, and `nixio.fs` precisely does what I did here: Have the base code be established by the C module and modify/extend it by Lua code. I have no idea why "making C metatables accessible always asks for trouble", and actually Lua does exactly that with its base types, doesn't it? But if that really is a bad idea, this patch obviously shouldn't be merged. Alternatively, I could implement the function in the C library completely, which would take quite some more LoC, but eliminate the need for another require and modifying metatables.

However, I would argue for having the factory as a method of the parser object, simply because that's the way luci-lib-json did it and it is the easiest way to bundle this IMO quite useful feature to the library.

jow- commented:
@jplitza We use a customized variant of LuaDoc. Try something like `./build/makedocs.sh documentation/api/` and you should see a suite of interlinked HTML files in the `documentation/api/` directory. I just ushed a few LuaDoc fixes recently so in doubt rebase your tree to the current HEAD.

NeoRaider commented:
When making the comment "making C metatables accessible always asks for trouble", I didn't see that the metatable is visible through `getmetatable` anyways and not hidden by a `__metatable` entry ("Progamming in Lua" recommends doing so to prevent changing `__gc`), so this is just adding a way to access the metatable without creating a parser instance. So no further concerns about this from my side.

jplitza commented:
I didn't like the hacky implementation of of the extension anymore myself, so I rewrote it in C. This eliminates the need for the metatable export and also for the additional `require`. Also, I signed the commit off and added LuaDoc. So in theory this should now be mergable, right? :)

karlp commented on libs/luci-lib-jsonc/src/jsonc.luadoc:
```diff
@@ -121,10 +121,22 @@ parser:set({ "some", "data" })`
+@usage `parser = luci.jsonc.new()
+ltn12.pump.all(ltn12.source.file(io.input()), parser:sink())
+print(parser:get())`
+]]
```
Might just be me, but that doesn't help me understand how to use it at all.

jplitza replied:
Do you know what ltn12 is? If not, no wonder you don't understand the documentation. ^^
I first had a `@see http://lua-users.org/wiki/FiltersSourcesAndSinks` line in there, but the LuaDoc doesn't cope well with URLs and tried to interpret it as a method name. So if you tell me how to integrate that link in a reasonable way, I'll happily include it.

karlp replied:
Maybe it's more that luci.jsonc is undocumented, so I don't see where this fits in.

jow- replied:
@karlp See http://htmlpreview.github.io/?https://github.com/openwrt/luci/blob/master/documentation/api/modules/luci.jsonc.html and https://htmlpreview.github.io/?https://raw.githubusercontent.com/openwrt/luci/master/documentation/api/modules/luci.jsonc.parser.html

karlp replied:
getting off topic, but where will this updated documentation end up?

jplitza commented:
I added a type check at the beginning of `json_parse_sink()` so the argument really is a parser object (although it would be hard to call it otherwise as the metatable isn't easily exposed).

So, what's the news? Apart from the documentation discussion (which seems to be more of an internal debate and has nothing to do with this particular pull request) I didn't see any more objections.

jow- commented:
Sorry for letting this linger for so long. I don't really have objections to it, so merged.
